File: youjing/frontend1111/js/hscode.js
const hscode_recordLoad = (evt_id, recordset) => {
    _.http.post('/api/saier/hscode/load/check_user', {}).then(res => {
        if (res.code != 1) {
            _.ui.message.error(res.msg || '加载校验失败')
            return
        }

        let is_hg_user = (res.data && res.data.is_hg_user == 1)

        if (is_hg_user) {
            recordset.module.field_by_full_name('海关编码').disabled = false
            recordset.module.field_by_full_name('退税率').disabled = false
            recordset.module.field_by_full_name('增值税率').disabled = false
            recordset.module.field_by_full_name('是否批准').disabled = false
            recordset.module.field_by_full_name('详细资料.是否批准').disabled = false
        } else {
            // 按 Pascal sb='1' 且非海关岗位逻辑：只禁用“是否批准”
            recordset.module.field_by_full_name('是否批准').disabled = true
            recordset.module.field_by_full_name('详细资料.是否批准').disabled = true
        }
    }).catch(err => {
        _.ui.message.error(err.msg || '请求失败')
    })
}

_.evts.on([_.evtids.RECORD_AFTER_EMPTY, _.evtids.RECORD_LOAD], hscode_recordLoad, '海关编码表')

const hscode_fieldChange = (evt_id, opts) => {
    let {
        recordset,
        module,
        field,
        value,
        old_value
    } = opts;
    let n = module.name

    if (field.full_name == n + '.海关编码') {
        _.http.post('/api/saier/hscode/hgbm/change', {
            '海关编码': recordset.val('海关编码'),
            '是否批准': recordset.val('是否批准'),
        }).then(res => {
            if (res.code != 1) {
                _.ui.message.error(res.msg || '处理失败')
                return
            }

            let fields = (res.data && res.data.fields) ? res.data.fields : {}
            Object.keys(fields).forEach(k => recordset.val(k, fields[k]))
        }).catch(err => {
            _.ui.message.error(err.msg || '请求失败')
        })
    }

    // 详细资料.是否批准.change
    if (field.full_name == n + '.详细资料.是否批准') {
        if (String(value || '') == '是') {
            let zzsl = Number(recordset.val('详细资料.增值税率') || 0)
            let tsl = Number(recordset.val('详细资料.退税率') || 0)

            // 子表逐行修改（不清空）
            let t = recordset.tables['详细资料']
            if (t && t.view_data) {
                for (let r of t.view_data) {
                    r.zzsl = zzsl
                    r.tsl = tsl
                    if (r.rid) t.push_modi_rid(r.rid)
                }
                t.sync_operate_data()
                t.modified = true
            }
            recordset.val('增值税率', zzsl)
            recordset.val('退税率', tsl)
        }
    }

    if (field.full_name == n + '.详细资料.退税率' ||
        field.full_name == n + '.详细资料.增值税率' ||
        field.full_name == n + '.详细资料.中文品名' ||
        field.full_name == n + '.详细资料.英文品名' ||
        field.full_name == n + '.详细资料.计量单位' ||
        field.full_name == n + '.详细资料.申报要素' ||
        field.full_name == n + '.详细资料.是否商检' ||
        field.full_name == n + '.详细资料.商检标志'
    ) {
        recordset.val('详细资料.新增人员', _.user.username)
        recordset.val('详细资料.是否批准', '否')
    }

    if (field.full_name == n + '.详细资料.中文品名' ||
        field.full_name == n + '.详细资料.申报要素' ||
        field.full_name == n + '.详细资料.计量单位'
    ) {
        let hwmc = recordset.val('详细资料.中文品名')
        let cznr = recordset.val('详细资料.申报要素')
        let hgbm = recordset.val('详细资料.海关编码')
        let sb = recordset.val('详细资料.识别')

        if (String(sb || '') == '' && String(hwmc || '') != '' && String(cznr || '') != '' && String(hgbm || '') != '') {
            _.http.post('/api/saier/hscode/detail/check_duplicate', {
                '中文品名': hwmc,
                '申报要素': cznr,
                '海关编码': hgbm,
                '识别': sb
            }).then(res => {
                if (res.code == 1 && res.data && res.data.duplicate == 1) {
                    _.ui.message.error('相同中文品名+报关要素的已有请检查!不能提交保存')
                    recordset.val('详细资料.是否批准', '重复')
                }
            }).catch(err => {
                _.ui.message.error(err.msg || '请求失败')
            })
        }
    }
}

_.evts.on(_.evtids.RECORD_FIELD_CHANGE, hscode_fieldChange, '海关编码表')

const hscode_table_new_after = (evt_id, table, recordset) => {
    if (table.group == '详细资料') {
        recordset.val('详细资料.退税率', recordset.val('退税率'))
        recordset.val('详细资料.增值税率', recordset.val('增值税率'))
        recordset.val('详细资料.海关编码', recordset.val('海关编码'))

        _.http.post('/api/saier/hscode/load/check_user', {}).then(res => {
            if (res.code == 1 && res.data && res.data.is_hg_user == 1) {
                recordset.val('详细资料.是否批准', '')
            }
        }).catch(err => {
            console.log(err)
            _.ui.message.error(err.msg)
        })
    }
}

_.evts.on([_.evtids.RECORD_TABLE_AFTER_NEW, _.evtids.RECORD_TABLE_AFTER_COPY],hscode_table_new_after,'海关编码表')

const hscode_before_save = (evt_id, recordset) => {
    return new Promise((resolve, reject) => {
        recordset.val('修改人员', _.user.username)

        let has_repeat = false
        const t = recordset.tables['详细资料']
        const details = (t && t.view_data) ? t.view_data : []

        for (let r of details) {
            r.hgbm = recordset.val('海关编码')
            r.tdhgbm = recordset.val('替代海关编码')
            r.jzcysb = recordset.val('禁止出运识别')
            r.jzbgsb = recordset.val('禁止报关识别')
            r.zzsl = Number(recordset.val('增值税率') || 0)
            r.tsl = Number(recordset.val('退税率') || 0)
            r.sfsj = recordset.val('是否商检')

            if (!r.sb) {
                r.sb = `${recordset.val('海关编码') || ''}/${r['中文品名'] || ''}/${r['申报要素'] || ''}${new Date().format('yyyy-MM-dd hh:mm:ss')}`
            }

            if (r.sfpz == '重复') {
                has_repeat = true
                _.ui.message.error(`中文品名:${r['中文品名'] || ''}已有请检查!不能提交保存`)
            }

            if (r.rid && t) t.push_modi_rid(r.rid)
        }

        if (t) {
            t.sync_operate_data()
            t.modified = true
        }

        if (has_repeat) {
            reject()
            return
        }

        _.http.post('/api/saier/hscode/before_save', {
            main: {
                hgbm: recordset.val('海关编码'),
                jzbgsb: recordset.val('禁止报关识别')
            },
            details: details
        }).then(res => {
            if (res.code == 1) resolve()
            else {
                _.ui.message.error(res.msg || '保存前处理失败')
                reject()
            }
        }).catch(err => {
            _.ui.message.error(err.msg || '请求失败')
            reject()
        })
    })
}

_.evts.on(_.evtids.RECORD_BEFORE_SAVE, hscode_before_save, '海关编码表')

File: youjing/youjing1111/hscode.py
from pdb import run
from any import *
from .model import *
from sqlalchemy.sql import func,not_,or_,and_
import time
from datetime import datetime
import json,os
from .__default__ import get_user_path,module_xxck_new,module_share_new,user_task_delete,user_task_new
    
SYS_FIELDS = ['sid','rid','uid','ctime','mtime','has_att','modi_uid','wf_status','wf_unit','pid','archived']

@any_route('/api/saier/hscode/load/check_user', methods=['POST', 'GET'])
@require_token
async def view_saier_hscode_load_check_user(request):
    s = Session()
    user = request.current_user
    try:
        username = str(user.username)
        rows = run_sql(
            f"select username from sys_user where username='{username}' and position like '%海关%' limit 1"
        )
        is_hg_user = 1 if len(rows) > 0 else 0
        return json_result(1, '查询成功', {'is_hg_user': is_hg_user})
    except:
        logger.error(trace_error())
        return json_result(-1, trace_error())
    finally:
        s.close()
        
@any_route('/api/saier/hscode/hgbm/change', methods=['POST', 'GET'])
@require_token
async def view_saier_hscode_hgbm_change(request):
    s = Session()
    user = request.current_user
    j = await request.json()
    try:
        hgbm1 = str(j.get('海关编码', ''))

        fields = {}

        if hgbm1 != '':
            d1 = run_sql(
                f"select rid from cyzglsheet "
                f"where '{hgbm1}' like concat(xm, '%') and zm='禁止出运海关编码' limit 1"
            )
            fields['禁止出运识别'] = '禁止出运' if len(d1) > 0 else '否'

            d2 = run_sql(
                f"select rid from cyzglsheet "
                f"where xm='{hgbm1}' and zm='禁止报关海关编码' limit 1"
            )
            fields['禁止报关识别'] = '禁止报关' if len(d2) > 0 else '否'
        else:
            fields['禁止出运识别'] = '否'
            fields['禁止报关识别'] = '否'

        return json_result(1, '处理成功', {'fields': fields})
    except:
        logger.error(trace_error())
        return json_result(-1, trace_error())
    finally:
        s.close()
        
@any_route('/api/saier/hscode/detail/check_duplicate', methods=['POST', 'GET'])
@require_token
async def view_saier_hscode_detail_check_duplicate(request):
    s = Session()
    user = request.current_user
    j = await request.json()
    try:
        hwmc = str(j.get('中文品名', ''))
        cznr = str(j.get('申报要素', ''))
        hgbm = str(j.get('海关编码', ''))
        sb = str(j.get('识别', ''))

        duplicate = 0
        if sb == '' and hwmc != '' and cznr != '' and hgbm != '':
            d = run_sql(
                f"select rid from hgbmbsheet "
                f"where hgbm='{hgbm}' and hwmc='{hwmc}' and cznr='{cznr}' limit 1"
            )
            if len(d) > 0:
                duplicate = 1

        return json_result(1, '处理成功', {'duplicate': duplicate})
    except:
        logger.error(trace_error())
        return json_result(-1, trace_error())
    finally:
        s.close()
        
@any_route('/api/saier/hscode/load/check_user', methods=['POST', 'GET'])
@require_token
async def view_saier_hscode_load_check_user(request):
    s = Session()
    user = request.current_user
    try:
        username = str(user.username)
        rows = run_sql(
            f"select username from sys_user where username='{username}' and position like '%海关%' limit 1"
        )
        return json_result(1, '查询成功', {'is_hg_user': 1 if len(rows) > 0 else 0})
    except:
        logger.error(trace_error())
        return json_result(-1, trace_error())
    finally:
        s.close()

@any_route('/api/saier/hscode/before_save', methods=['POST', 'GET'])
@require_token
async def view_saier_hscode_before_save(request):
    s = Session()
    user = request.current_user
    j = await request.json()
    try:
        main = j.get('main', {})
        details = j.get('details', [])

        hgbm1 = str(main.get('hgbm', ''))
        jzbgsb = str(main.get('jzbgsb', ''))
        now_dt = time.strftime("%Y-%m-%d %H:%M:%S")
        now_day = time.strftime("%Y-%m-%d")

        # 禁止报关海关编码 -> cyzglsheet
        if hgbm1 != '' and jzbgsb == '禁止报关':
            row = s.query(cyzglsheet).filter(
                cyzglsheet.xm == hgbm1,
                cyzglsheet.zm == '禁止报关海关编码'
            ).first()
            if not row:
                parent_row = s.query(cyzgl).filter(cyzgl.zm == '禁止报关海关编码').first()
                if parent_row:
                    m = cyzglsheet()
                    m.rid = get_uuid()
                    m.uid = user.rid
                    m.ctime = now_dt
                    m.mtime = now_dt
                    m.pid = str(parent_row.rid)
                    m.zm = '禁止报关海关编码'
                    m.xm = hgbm1
                    m.qxzl = '禁止报关海关编码'
                    s.add(m)

        for r in details:
            sfpz = str(r.get('sfpz', ''))
            if sfpz == '重复':
                s.rollback()
                return json_result(-1, f"中文品名:{str(r.get('中文品名', ''))}已有请检查!不能提交保存")

            # 否：发消息/任务（不处理 ddsp）
            if sfpz == '否':
                xzry = str(r.get('xzry', ''))
                mgr = ''
                mgr_row = s.query(sys_user).filter(sys_user.username == xzry).first()
                if mgr_row and getattr(mgr_row, 'mgr', ''):
                    mgr = str(mgr_row.mgr)

                if mgr != '' and mgr != user.username:
                    xxnr = f"{user.username}的海关编码:{hgbm1}申请请查看:{now_day}"
                    xx_row = {
                        'sys_owner': mgr,
                        'sys_path': '我的公司\\宁波优景进出口有限公司\\',
                        'fsrq': now_day,
                        'fsr': user.username,
                        'xxly': '海关编码',
                        'bjdh': '',
                        'wxht': hgbm1,
                        'cght': '',
                        'gdht': '',
                        'yhdh': '',
                        'xxnr': xxnr,
                        'jsr': mgr
                    }
                    res = module_xxck_new([xx_row], user, s)
                    if res.get('code') != 1:
                        s.rollback()
                        return json_result(res.get('code'), res.get('msg'))

            # 是：同步税率 + hgbm维护
            if sfpz == '是':
                hgbm_val = str(r.get('hgbm', ''))
                hwmc = str(r.get('hwmc', ''))
                ywpm = str(r.get('ywpm', ''))
                cznr = str(r.get('cznr', ''))
                dw = str(r.get('dw', ''))
                sfsj = str(r.get('sfsj', ''))
                sjbz = str(r.get('sjbz', ''))
                tsl = float(r.get('tsl', 0) or 0)
                zzsl = float(r.get('zzsl', 0) or 0)

                s.query(cjcp).filter(cjcp.hgbm == hgbm_val).update({'tsl': tsl, 'zzsl': zzsl})
                s.query(zscp).filter(zscp.hgbm == hgbm_val).update({'tsl': tsl, 'zzsl': zzsl})

                one = s.query(hgbm).filter(
                    hgbm.hgbm == hgbm_val,
                    hgbm.hwmc == hwmc,
                    hgbm.cznr == cznr
                ).first()

                if one:
                    s.query(hgbm).filter(hgbm.rid == str(one.rid)).update({
                        'hgbm': hgbm1,
                        'tsl': tsl,
                        'zzsl': zzsl,
                        'ywpm': ywpm,
                        'dw': dw,
                        'sfsj': sfsj,
                        'sjbz': sjbz,
                        'cznr': cznr
                    })
                else:
                    ms = hgbm()
                    ms.rid = get_uuid()
                    ms.uid = user.rid
                    ms.ctime = now_dt
                    ms.mtime = now_dt
                    ms.hgbm = hgbm1
                    ms.hwmc = hwmc
                    ms.tsl = tsl
                    ms.zzsl = zzsl
                    ms.ywpm = ywpm
                    ms.dw = dw
                    ms.sfsj = sfsj
                    ms.sjbz = sjbz
                    ms.cznr = cznr
                    s.add(ms)

        s.commit()
        return json_result(1, '处理成功')
    except:
        s.rollback()
        logger.error(trace_error())
        return json_result(-1, trace_error())
    finally:
        s.close()